dvc/repo/index.py

```diff
@@ -0,0 +1,285 @@
+        new_stages = set(stages)
+        # we remove existing stages with same hashes at first
+        # and then re-add the new ones later.
+        stages_set = (set(self.stages) - new_stages) | new_stages
+        return Index(self.repo, self.fs, stages=list(stages_set))
+
+    def add(self, stage: "Stage") -> "Index":
+        return self.update([stage])
+
+    def remove(
+        self, stage: "Stage", ignore_not_existing: bool = False
+    ) -> "Index":
+        stages = self._discard_stage(
+            stage, ignore_not_existing=ignore_not_existing
+        )
+        return Index(self.repo, self.fs, stages=stages)
+
+    def discard(self, stage: "Stage") -> "Index":
+        return self.remove(stage, ignore_not_existing=True)
+
+    def difference(self, stages: Iterable["Stage"]) -> "Index":
+        # this does not preserve the order
+        stages_set = set(self.stages) - set(stages)
+        return Index(self.repo, self.fs, stages=list(stages_set))
+
+    def _discard_stage(
+        self, stage: "Stage", ignore_not_existing: bool = False
+    ) -> List["Stage"]:
+        stages = self.stages[:]
+        ctx = suppress(ValueError) if ignore_not_existing else nullcontext()
+        with ctx:
+            stages.remove(stage)
+        return stages
+
+    @memoize
+    def build_graph(self) -> "DiGraph":
+        from dvc.repo.graph import build_graph
+
+        return build_graph(self.stages, self.outs_trie)
+
+    def check_graph(self) -> None:
+        if not getattr(self.repo, "_skip_graph_checks", False):
+            self.build_graph()
+
+    def dumpd(self) -> Dict[str, Dict]:
+        def dump(stage: "Stage"):
+            key = stage.path_in_repo
+            try:
+                key += ":" + stage.name  # type: ignore[attr-defined]
+            except AttributeError:
+                pass
+            return key, stage.dumpd()
+
+        return dict(dump(stage) for stage in self)
+
+    @cached_property
+    def identifier(self) -> str:
+        """Unique identifier for the index.
+
+        We can use this to optimize and skip opening some indices
+        eg: on push/pull/fetch/gc --all-commits.
+
+        Currently, it is unique to the platform (windows vs posix).
+        """
+        return dict_md5(self.dumpd())
+
+
+if __name__ == "__main__":
+    from funcy import log_durations
+
+    from dvc.repo import Repo
+
+    repo = Repo()
+    index = Index(repo, repo.fs)
+    print(index)
+    with log_durations(print, "collecting stages"):
+        # pylint: disable=pointless-statement
+        print("no of stages", len(index.stages))
+    with log_durations(print, "building graph"):
+        index.build_graph()
+    with log_durations(print, "calculating hash"):
+        print(index.identifier)
+    with log_durations(print, "updating"):
+        index2 = index.update(index.stages)
+    with log_durations(print, "calculating hash"):
+        print(index2.identifier)
```

Comment:
```suggestion
```

Reply:
I think I'll keep these as it is quite helpful in checking bugs/performance issues quickly, unless there is some strong reason against this.

Reply:
@skshetry This is not suitable for the core codebase, just create a test instead.

Reply:
@skshetry Looks like we also still have these in parsing and in ui. They really should not be there.

Reply:
But I cannot run test on a real project, with this, I can do that by just `python dvc/repo/index.py`.

Reply:
> They really should not be there.

Could you please elaborate in why you don't like this?

I really like having small snippets as it allows me to document  example usages, play with and test quite easily.

Other projects also use them in a similar manner, eg: [rich](https://github.com/willmcgugan/rich/blob/542cfe1a9681569123366d51d5bb2e53eef8c62c/rich/table.py#L845).

Reply:
> no one is using it except you

Right, because it's not merged yet. Like I said above, it acts as a quick litmus test, especially for UI where I can quickly visually inspect if there's any issues.

> Keeping such snippets around is very ugly.

That's quite subjective, it's not a utility at all.

Reply:
Oops, sorry, didn't see your last comment.

> Could you please elaborate in why you don't like this?
> I really like having small snippets as it allows me to document example usages, play with and test quite easily.

It is a test script that is not used anywhere in the code or during normal workflow. It is also not a generic debugging mechanism like --pdb/--cprofile/etc. It looks to me like an arbitrary test code that no one knows about, which makes the codebase smell, as it is percieved like some debugging code that we forgot to remove.

> Other projects also use them in a similar manner, eg: rich.

It might be more justifiable for rich, because it is a library, but it still looks to me as a code snippet that someone forgot to remove.

Reply:
> Right, because it's not merged yet. Like I said above, it acts as a quick litmus test, especially for UI where I can quickly visually inspect if there's any issues.

The functionality makes sense to me - to quickly check if our subsystems work correctly. I only have a problem with code snippets scattered around the codebase, that no one even knows exist and could be ran. How about we create a command (`dvc test`? `dvc doctor`? `dvc debug`?), where we will collect all such self-checks, so that there is one place to test if dvc subsystems are operating correctly?

Reply:
For having a central way of running all these, it requires certain pre-conditions/post-conditions (eg: this needs to always be run inside a DVC repo). And, having a central way is not the point of this, tests: end-to-end or unit is there for that.

> but it still looks to me as a code snippet that someone forgot to remove.

It is very much intentional. They use it for showing example usages and demonstration, not to add quick visual inspection.

> no one even knows exist and could be ran

It's okay to not know they exist, as they are very local to the subsystem. Whenever someone happens to work in this module, they'll take a note of `if __name__ == "main"`. When refactoring, they'll be able to grep it or find usages of it to fix. And, as it's not critical to us, it's okay if it breaks occasionally. Though the parsing/UI has never been broken, they still work.

>  one place to test

The parsing one is a test and uses doctest, so we could run it always. But there is not much benefit on running others, as they serve different purpose.

Reply:
@skshetry Are there any other examples besides rich? I feel like it is a very specific library, I would love to see some other examples in serious projects (cli tools would be great). Just trying to understand when and how this is used.

CC @iterative/dvc , guys, wdyt about this practice?

Reply:
I'd completely agree with moving this to `tests/`

Reply:
Ok, so looks like this was indeed a practice that wasn't aware of. Thanks to @isidentical and @skshetry for pointing out that this is used in things like cpython, doctests, rich. However, I wasn't able to find a single case in any of the projects that I've looked through (e.g. numpy, pyinstaller, duliwch, boto, moto, scipy, spacy, etc), so this still feels like a pretty cryptic practice to me, so I'm not personally 100% sold on it, it feels like these snippets belong somewhere else. But there doesn't seem to be much opposition to it and we already have a few of these in the codebase, so let's keep it for now, and get back to discussing this practice in general on retro or elsewhere.